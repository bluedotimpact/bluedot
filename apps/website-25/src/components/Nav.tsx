import clsx from 'clsx';
import React, {
  useCallback, useState, useEffect,
} from 'react';
import ClickAwayListener from 'react-click-away-listener';

import { CTALinkOrButton } from '@bluedot/ui/src/CTALinkOrButton';
import { HamburgerButton } from '@bluedot/ui/src/HamburgerButton';
import { ROUTES } from '../lib/routes';

export type NavProps = React.PropsWithChildren<{
  className?: string;
  logo?: string;
  courses: Array<{
    title: string;
    href: string;
    isNew?: boolean;
  }>;
}>;

const DropdownIcon: React.FC<{ expanded: boolean }> = ({ expanded }) => (
  <svg
    className={clsx('size-4 transition-transform', expanded ? 'rotate-180' : '')}
    viewBox="0 0 24 24"
    fill="none"
    stroke="currentColor"
  >
    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
  </svg>
);

const ExploreSection: React.FC<{
  expanded: boolean;
  innerClassName?: string;
  className?: string;
  courses: Array<{ title: string; href: string; isNew?: boolean }>;
  isScrolled?: boolean;
}> = ({
  expanded, innerClassName, className, courses, isScrolled,
}) => (
  <div
    className={clsx(
      'nav-explore-section__content-wrapper overflow-hidden transition-[max-height,opacity] duration-300',
      expanded ? 'max-h-96 opacity-100' : 'max-h-0 opacity-0',
      isScrolled && '[&_a]:link-on-dark',
      className,
    )}
  >
    <div
      className={clsx(
        'nav-explore-section___dropdown-content flex flex-col gap-[14px] text-pretty px-6',
        innerClassName,
      )}
    >
      <h3 className="nav-explore-section__dropdown-title font-bold">Our courses</h3>
      {courses?.map((course) => (
        <a key={course.href} href={course.href} tabIndex={expanded ? 0 : -1} className="nav-explore-section__dropdown-link">
          {course.isNew && (
            <span className="nav-explore-section__new-badge text-bluedot-normal font-black pr-2">
              New!
            </span>
          )}
          {course.title}
        </a>
      ))}
    </div>
  </div>
);

const NavLinks: React.FC<{
  exploreSectionInline?: boolean;
  courses: Array<{ title: string; href: string; isNew?: boolean }>;
  exploreExpanded: boolean;
  onToggleExplore: () => void;
  className?: string;
  isScrolled: boolean;
}> = ({
  exploreSectionInline,
  courses,
  exploreExpanded,
  onToggleExplore,
  className,
  isScrolled,
}) => {
  const navLinkContainerClasses = clsx(
    'nav-links grid [&>*]:w-fit items-center',
    exploreSectionInline ? 'grid-cols-1 gap-y-9' : 'grid-cols-subgrid grid-rows-subgrid row-[1/3] col-[3/7]',
    className,
  );
  const navLinkClasses = clsx('nav-link-animation', isScrolled && 'nav-link-animation-dark');

  const exploreSection = (
    <ExploreSection
      expanded={exploreExpanded}
      courses={courses}
      className="nav-links__explore-section col-span-full row-[2]"
      innerClassName={exploreSectionInline ? 'pt-6' : 'pb-9'}
      isScrolled={isScrolled}
    />
  );

  return (
    <div className={navLinkContainerClasses}>
      <div>
        <button
          type="button"
          onClick={onToggleExplore}
          className={clsx('nav-links__dropdown-button flex items-center gap-2', navLinkClasses)}
        >
          Explore
          <DropdownIcon expanded={exploreExpanded} />
        </button>
        {exploreSectionInline && exploreSection}
      </div>
      {/* `exploreSection` is placed here in the DOM to make it the next step after "Explore"
          when tabbing through the header. Visually, it is moved below by setting row-[2] */}
      {!exploreSectionInline && exploreSection}
      <a href={ROUTES.about.url} className={clsx('nav-links__link', navLinkClasses, isCurrentPath(ROUTES.about.url) && 'font-bold')}>About us</a>
      <a href={ROUTES.joinUs.url} className={clsx('nav-links__link', navLinkClasses, isCurrentPath(ROUTES.joinUs.url) && 'font-bold')}>Join us</a>
      <a href="https://bluedot.org/blog/" className={clsx('nav-links__link', navLinkClasses)}>Blog</a>
    </div>
  );
};

const CTAButtons: React.FC<{ className?: string }> = ({ className }) => (
  <div className={clsx('nav__cta-container', className)}>
    <CTALinkOrButton
      className="nav__primary-cta"
      url="https://donate.stripe.com/5kA3fpgjpdJv6o89AA"
      isExternalUrl
    >
      Support us
    </CTALinkOrButton>
    {/*
      TODO Potentially switch from "Support us" to "Log in"/"Get started for free" for non-MVP site
      <a className="nav__secondary-cta" href="https://course.aisafetyfundamentals.com/alignment?show=login" {...EXTERNAL_LINK_PROPS}>Log in</a>
      <CTALinkOrButton className="nav__primary-cta" url="https://aisafetyfundamentals.com/" {...EXTERNAL_LINK_PROPS}>Get started for free</CTALinkOrButton>
    */}
  </div>
);

export const isCurrentPath = (href: string): boolean => {
  if (typeof window === 'undefined') return false;
  const currentPath = window.location.pathname;
  return href === currentPath || (href !== '/' && currentPath.startsWith(href));
};

export const Nav: React.FC<NavProps> = ({
  className, logo, courses,
}) => {
  const [expandedSections, setExpandedSections] = useState<'none' | 'nav' | 'nav-and-explore'>('none');
  const navExpanded = expandedSections === 'nav' || expandedSections === 'nav-and-explore';
  const exploreExpanded = expandedSections === 'nav-and-explore';

  const [isScrolled, setIsScrolled] = useState(false);

  const onToggleExplore = useCallback(() => {
    setExpandedSections((prev) => (prev === 'nav-and-explore' ? 'nav' : 'nav-and-explore'));
  }, []);

  const onToggleNav = useCallback(() => {
    setExpandedSections((prev) => (prev === 'none' ? 'nav' : 'none'));
  }, []);

  useEffect(() => {
    const handleScroll = () => {
      setIsScrolled(window.scrollY > 0);
    };

    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, []);

  return (
    <nav
      className={clsx(
        'nav fixed z-50 w-full container-elevated transition-all duration-300',
        isScrolled ? 'bg-color-canvas-dark **:text-white' : 'bg-color-canvas',
        className,
      )}
    >
      <ClickAwayListener onClickAway={() => setExpandedSections('none')}>
        <div className="nav__container section-base">
          <div className="nav__bar w-full grid grid-cols-2 gap-x-9 lg:grid-cols-[20%_1fr_auto_auto_auto_auto_1fr_20%] items-center grid-rows-[72px] sm:grid-rows-[100px]">
            <a href="/" className="nav__logo-link shrink-0 w-[200px] col-[1]">
              {logo ? (
                <img
                  className={clsx(
                    'nav__logo h-6 mr-auto transition-all duration-300',
                    isScrolled && 'brightness-0 invert',
                  )}
                  src={logo}
                  alt="BlueDot Impact Logo"
                />
              ) : (
                <h3 className="nav_logo--placeholder h-8 mr-auto">BlueDot Impact</h3>
              )}
            </a>
            <NavLinks
              onToggleExplore={onToggleExplore}
              exploreExpanded={exploreExpanded}
              courses={courses}
              isScrolled={isScrolled}
              className="nav__links--desktop hidden lg:grid mx-auto"
            />
            <div className="nav__actions flex gap-space-between ml-auto col-[8]">
              <CTAButtons className="nav__login--tablet-desktop gap-6 hidden sm:flex" />
              <HamburgerButton
                open={navExpanded}
                setOpen={onToggleNav}
                className="nav__menu--mobile-tablet lg:hidden"
              />
            </div>
          </div>

          <div
            className={clsx(
              'nav__drawer transition-[max-height,opacity] duration-300 relative overflow-hidden px-3',
              navExpanded ? 'max-h-[800px] opacity-100' : 'max-h-0 opacity-0',
            )}
          >
            <div className="nav__drawer-content--mobile-tablet flex flex-col grow font-medium pb-8 pt-2 lg:hidden">
              <NavLinks
                onToggleExplore={onToggleExplore}
                exploreExpanded={exploreExpanded}
                exploreSectionInline
                courses={courses}
                isScrolled={isScrolled}
                className="nav__links--mobile-tablet flex-col"
              />
              <CTAButtons className="nav__login--mobile justify-between mt-20 flex sm:hidden" />
            </div>
          </div>
        </div>
      </ClickAwayListener>
    </nav>
  );
};
