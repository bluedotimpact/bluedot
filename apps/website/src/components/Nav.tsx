import clsx from 'clsx';
import React, { useState, useEffect } from 'react';
import { FaCircleUser } from 'react-icons/fa6';
import ClickAwayListener from 'react-click-away-listener';

import { CTALinkOrButton } from '@bluedot/ui/src/CTALinkOrButton';
import { IconButton, HamburgerIcon } from '@bluedot/ui/src/IconButton';
import { useAuthStore, addQueryParam } from '@bluedot/ui';
import { ROUTES } from '../lib/routes';
import { A, H3 } from './Text';

export type NavProps = React.PropsWithChildren<{
  className?: string;
  logo?: string;
  primaryCtaText?: string;
  primaryCtaUrl?: string;
  courses: {
    title: string;
    url: string;
    isNew?: boolean;
  }[];
}>;

const TRANSITION_DURATION_CLASS = 'duration-300';

const DropdownIcon: React.FC<{ expanded: boolean }> = ({ expanded }) => (
  <svg
    className={clsx('size-4 transition-transform', expanded ? 'rotate-180' : '')}
    viewBox="0 0 24 24"
    fill="none"
    stroke="currentColor"
  >
    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
  </svg>
);

const ExploreSection: React.FC<{
  expanded: boolean;
  innerClassName?: string;
  className?: string;
  courses: { title: string; url: string; isNew?: boolean }[];
  isScrolled?: boolean;
}> = ({
  expanded, innerClassName, className, courses, isScrolled,
}) => (
  <div
    className={clsx(
      `nav-explore-section__content-wrapper overflow-hidden transition-[max-height,opacity] ${TRANSITION_DURATION_CLASS}`,
      expanded ? 'max-h-96 opacity-100' : 'max-h-0 opacity-0',
      isScrolled && '[&_a]:link-on-dark',
      className,
    )}
  >
    <div
      className={clsx(
        'nav-explore-section___dropdown-content flex flex-col gap-[14px] w-fit overflow-hidden text-pretty',
        innerClassName,
      )}
    >
      <H3 className="nav-explore-section__dropdown-title font-bold">Our courses</H3>
      {courses?.map((course) => (
        <A key={course.url} href={course.url} className="nav-explore-section__dropdown-link no-underline">
          {course.isNew && (
            <span className="nav-explore-section__new-badge text-bluedot-normal font-black pr-2">
              New!
            </span>
          )}
          {course.title}
        </A>
      ))}
    </div>
  </div>
);

const NavLinks: React.FC<{
  exploreSectionInline?: boolean;
  courses: { title: string; url: string; isNew?: boolean }[];
  exploreExpanded: boolean;
  onToggleExplore: () => void;
  className?: string;
  isScrolled: boolean;
}> = ({
  exploreSectionInline,
  courses,
  exploreExpanded,
  onToggleExplore,
  className,
  isScrolled,
}) => {
  const navLinkClasses = clsx('nav-link-animation no-underline', isScrolled && 'nav-link-animation-dark');

  return (
    <div className={clsx('nav-links flex gap-9 [&>*]:w-fit', className)}>
      <div>
        <button
          type="button"
          onClick={onToggleExplore}
          className={clsx('nav-links__dropdown-button flex items-center gap-2 cursor-pointer', navLinkClasses)}
        >
          Explore
          <DropdownIcon expanded={exploreExpanded} />
        </button>
        {exploreSectionInline && (
          <ExploreSection
            expanded={exploreExpanded}
            courses={courses}
            className="nav-links__explore-section"
            innerClassName="pl-6 pt-6"
            isScrolled={isScrolled}
          />
        )}
      </div>
      <A href={ROUTES.about.url} className={clsx('nav-links__link', navLinkClasses, isCurrentPath(ROUTES.about.url) && 'font-bold')}>About us</A>
      <A href={ROUTES.joinUs.url} className={clsx('nav-links__link', navLinkClasses, isCurrentPath(ROUTES.joinUs.url) && 'font-bold')}>Join us</A>
      <A href={ROUTES.blog.url} className={clsx('nav-links__link', navLinkClasses, isCurrentPath(ROUTES.blog.url) && 'font-bold')}>Blog</A>
      <A href="https://lu.ma/aisafetycommunityevents?utm_source=website&utm_campaign=nav" className={clsx('nav-links__link', navLinkClasses)}>Events</A>
    </div>
  );
};

const ProfileLinks: React.FC<{ isScrolled: boolean }> = ({ isScrolled }) => {
  const linkClasses = clsx('nav-link-animation w-fit no-underline', isScrolled && 'nav-link-animation-dark');

  return (
    <div className="nav__profile-dropdown flex flex-col gap-4 font-medium pb-10 items-end">
      <A href={ROUTES.profile.url} className={clsx('nav__profile-link', linkClasses)}>Profile</A>
      <A href={ROUTES.contact.url} className={clsx('nav__profile-link', linkClasses)}>Help</A>
      <A href={ROUTES.logout.url} className={clsx('nav__profile-link', linkClasses)}>Log out</A>
    </div>
  );
};

const CTAButtons: React.FC<{
  className?: string;
  primaryCtaText?: string;
  primaryCtaUrl?: string;
  isLoggedIn: boolean;
}> = ({
  className,
  primaryCtaText = 'Support us',
  primaryCtaUrl = 'https://donate.stripe.com/5kA3fpgjpdJv6o89AA',
  isLoggedIn,
}) => {
  const [loginUrl, setLoginUrl] = useState(ROUTES.login.url);
  useEffect(() => {
    if (typeof window !== 'undefined') {
      setLoginUrl(addQueryParam(ROUTES.login.url, 'redirect_to', window.location.pathname));
    }
  }, []);

  return (
    <div className={clsx('nav__cta-container', className)}>
      {!isLoggedIn && (
        <CTALinkOrButton
          className="nav__secondary-cta"
          variant="secondary"
          url={loginUrl}
        >
          Login
        </CTALinkOrButton>
      )}
      <CTALinkOrButton
        className="nav__primary-cta"
        variant="primary"
        url={primaryCtaUrl}
      >
        {primaryCtaText}
      </CTALinkOrButton>
    </div>
  );
};

type ExpandedSectionsState = {
  mobileNav: boolean;
  explore: boolean;
  profile: boolean;
};

export const isCurrentPath = (url: string): boolean => {
  if (typeof window === 'undefined') return false;
  const currentPath = window.location.pathname;
  return url === currentPath || (url !== '/' && currentPath.startsWith(url));
};

export const Nav: React.FC<NavProps> = ({
  className, logo, courses, primaryCtaText, primaryCtaUrl,
}) => {
  const [expandedSections, setExpandedSections] = useState<ExpandedSectionsState>({
    mobileNav: false,
    explore: false,
    profile: false,
  });

  const anyDrawerOpen = Object.values(expandedSections).some((v) => v);

  const isLoggedIn = !!useAuthStore((s) => s.auth);
  const [isScrolled, setIsScrolled] = useState(false);

  const onToggleExplore = () => setExpandedSections((prev) => ({
    mobileNav: prev.mobileNav,
    explore: !prev.explore,
    profile: false,
  }));

  const onToggleNav = () => setExpandedSections((prev) => ({
    mobileNav: !prev.mobileNav,
    explore: false,
    profile: false,
  }));

  const onToggleProfile = () => setExpandedSections((prev) => ({
    mobileNav: false,
    explore: false,
    profile: !prev.profile,
  }));

  useEffect(() => {
    const handleScroll = () => {
      setIsScrolled(window.scrollY > 0);
    };

    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, []);

  const drawerBaseClassName = 'transition-[max-height,opacity] relative overflow-hidden px-3';
  const drawerClosedClassName = `max-h-0 opacity-0 ${
    anyDrawerOpen ? 'duration-0' : TRANSITION_DURATION_CLASS
  }`; // Close instantly if any other drawer has been opened
  const drawerOpenClassName = `max-h-[700px] opacity-100 ${TRANSITION_DURATION_CLASS}`;

  return (
    <nav
      className={clsx(
        `nav sticky top-0 z-50 w-full container-elevated transition-all ${TRANSITION_DURATION_CLASS}`,
        isScrolled ? 'bg-color-canvas-dark **:text-white' : 'bg-color-canvas',
        className,
      )}
    >
      <ClickAwayListener onClickAway={() => setExpandedSections({ mobileNav: false, explore: false, profile: false })}>
        <div className="nav__container section-base">
          <div className="nav__bar w-full flex justify-between lg:grid lg:grid-cols-[20%_60%_20%] items-center h-[72px] sm:h-[100px]">
            <div className="flex gap-space-between items-center">
              <IconButton
                open={expandedSections.mobileNav}
                Icon={<HamburgerIcon />}
                setOpen={onToggleNav}
                className="nav__menu--mobile-tablet mr-2 lg:hidden"
              />
              <A href="/" className="nav__logo-link shrink-0 w-[200px] no-underline">
                {logo ? (
                  <img
                    className={clsx(
                      `nav__logo h-6 mr-auto transition-all ${TRANSITION_DURATION_CLASS}`,
                      isScrolled && 'brightness-0 invert',
                    )}
                    src={logo}
                    alt="BlueDot Impact Logo"
                  />
                ) : (
                  <H3 className="nav_logo--placeholder h-8 mr-auto">BlueDot Impact</H3>
                )}
              </A>
            </div>
            <NavLinks
              onToggleExplore={onToggleExplore}
              exploreExpanded={expandedSections.explore}
              courses={courses}
              isScrolled={isScrolled}
              className="nav__links--desktop hidden lg:flex mx-auto"
            />
            <div className="nav__actions flex gap-space-between ml-auto items-center">
              <CTAButtons
                primaryCtaText={primaryCtaText}
                primaryCtaUrl={primaryCtaUrl}
                isLoggedIn={isLoggedIn}
                className="nav__login--tablet-desktop gap-6 hidden sm:flex"
              />
              {isLoggedIn ? (
                <IconButton
                  open={expandedSections.profile}
                  Icon={<FaCircleUser className="size-[24px] opacity-75" />}
                  setOpen={onToggleProfile}
                  className="nav__profile-menu ml-2"
                />
              ) : (
                <CTALinkOrButton
                className="nav__login--mobile block sm:hidden"
                variant="secondary"
                url={ROUTES.login.url}
              >
                Login
              </CTALinkOrButton>
              )}
            </div>
          </div>
          <ExploreSection
            expanded={expandedSections.explore}
            courses={courses}
            className="nav__drawer-content--desktop"
            innerClassName="pb-10 hidden lg:flex mx-auto"
            isScrolled={isScrolled}
          />
          <div className={clsx('nav__links-drawer', drawerBaseClassName, expandedSections.mobileNav ? drawerOpenClassName : drawerClosedClassName)}>
            {/* Mobile & Tablet content (including Explore) */}
            <div className="nav__drawer-content--mobile-tablet flex flex-col grow font-medium pb-8 pt-2 lg:hidden">
              <NavLinks
                onToggleExplore={onToggleExplore}
                exploreExpanded={expandedSections.explore}
                exploreSectionInline
                courses={courses}
                isScrolled={isScrolled}
                className="nav__links--mobile-tablet flex-col"
              />
              <CTAButtons
                primaryCtaText={primaryCtaText}
                primaryCtaUrl={primaryCtaUrl}
                className="nav__login--mobile justify-end gap-6 mt-20 flex sm:hidden"
                isLoggedIn={isLoggedIn}
              />
            </div>
          </div>
          <div className={clsx('nav__profile-drawer', drawerBaseClassName, expandedSections.profile ? drawerOpenClassName : drawerClosedClassName)}>
            <ProfileLinks isScrolled={isScrolled} />
          </div>
        </div>
      </ClickAwayListener>
    </nav>
  );
};
