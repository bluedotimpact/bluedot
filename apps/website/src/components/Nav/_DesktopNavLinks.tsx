import { NavLinks } from './_NavLinks';
import { ExpandedSectionsState } from './utils';

export const DesktopNavLinks: React.FC<{
  expandedSections: ExpandedSectionsState;
  updateExpandedSections: (updates: Partial<ExpandedSectionsState>) => void;
  isScrolled: boolean;
}> = ({
  expandedSections,
  updateExpandedSections,
  isScrolled,
}) => {
  return (
    <NavLinks
      expandedSections={expandedSections}
      updateExpandedSections={updateExpandedSections}
      isScrolled={isScrolled}
      className="hidden lg:flex"
    />
  );
};
