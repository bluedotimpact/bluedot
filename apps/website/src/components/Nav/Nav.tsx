import clsx from 'clsx';
import React, { useState, useEffect } from 'react';
import ClickAwayListener from 'react-click-away-listener';
import { useAuthStore } from '@bluedot/ui';

import { NavLogo } from './_NavLogo';
import { NavCta } from './_NavCta';
import { MobileNavLinks } from './_MobileNavLinks';
import { DesktopNavLinks } from './_DesktopNavLinks';
import { ExpandedSectionsState, TRANSITION_DURATION_CLASS } from './utils';

export type NavProps = {
  className?: string;
  logo?: string;
};

export const Nav: React.FC<NavProps> = ({
  className,
  logo,
}) => {
  const isLoggedIn = !!useAuthStore((s) => s.auth);
  const [isScrolled, setIsScrolled] = useState(false);

  const [expandedSections, setExpandedSections] = useState<ExpandedSectionsState>({
    about: false,
    explore: false,
    mobileNav: false,
    profile: false,
  });

  const updateExpandedSections = (updates: Partial<ExpandedSectionsState>) => {
    setExpandedSections((prev: ExpandedSectionsState) => ({ ...prev, ...updates }));
  };

  useEffect(() => {
    const handleScroll = () => {
      setIsScrolled(window.scrollY > 0);
    };

    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, []);

  return (
    <nav
      className={clsx(
        `nav sticky top-0 z-50 w-full transition-all ${TRANSITION_DURATION_CLASS}`,
        isScrolled ? 'bg-color-canvas-dark **:text-white' : 'bg-white border-b border-color-divider',
        className,
      )}
    >
      <ClickAwayListener onClickAway={() => setExpandedSections({
        about: false,
        explore: false,
        mobileNav: false,
        profile: false,
      })}
      >
        <div className="nav__container section-base">
          <div className="nav__bar w-full flex justify-between items-center h-16">
            {/* Mobile & Tablet: Hamburger Button */}
            <MobileNavLinks
              expandedSections={expandedSections}
              updateExpandedSections={updateExpandedSections}
              isScrolled={isScrolled}
            />

            {/* Logo */}
            <NavLogo logo={logo} isScrolled={isScrolled} />

            {/* Desktop: Nav Links */}
            <DesktopNavLinks
              expandedSections={expandedSections}
              updateExpandedSections={updateExpandedSections}
              isScrolled={isScrolled}
            />

            {/* CTA Buttons */}
            <NavCta
              isLoggedIn={isLoggedIn}
              isScrolled={isScrolled}
              expandedSections={expandedSections}
              updateExpandedSections={updateExpandedSections}
            />
          </div>
        </div>
      </ClickAwayListener>
    </nav>
  );
};

export default Nav;
