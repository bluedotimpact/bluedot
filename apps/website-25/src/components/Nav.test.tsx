import {
  describe, expect, test,
  vi,
} from 'vitest';
import { render, waitFor, fireEvent } from '@testing-library/react';
import { Nav } from './Nav';

const mockLoggedInUser = () => {
  vi.mock('@bluedot/ui', async () => {
    const actual = await vi.importActual('@bluedot/ui');
    return {
      ...actual,
      useAuthStore: vi.fn().mockImplementation(() => ({
        auth: { token: 'mockToken', expiresAt: Date.now() + 10000 },
        setAuth: vi.fn(),
        internal_clearTimer: null,
        internal_refreshTimer: null,
      })),
    };
  });
};

describe('Nav', () => {
  test('renders with courses', () => {
    const { container } = render(
      <Nav
        logo="logo.png"
        courses={[
          { title: 'Course 1', url: '/course1' },
          { title: 'Course 2', url: '/course2', isNew: true },
        ]}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  test('clicking the hamburger button expands the mobile nav drawer', async () => {
    const { container } = render(
      <Nav courses={[{ title: 'Course 1', url: '/course1' }]} />,
    );

    const hamburgerButton = container.querySelector('.nav__menu--mobile-tablet');
    expect(hamburgerButton).not.toBeNull();

    const mobileNavDrawer = container.querySelector('.nav__links-drawer');
    const profileDrawer = container.querySelector('.nav__profile-drawer');
    expect(mobileNavDrawer).not.toBeNull();
    expect(profileDrawer).not.toBeNull();

    // Initially, both drawers should have a max height of 0 (closed state).
    expect(mobileNavDrawer!.className).toMatch(/max-h-0/);
    expect(profileDrawer!.className).toMatch(/max-h-0/);

    fireEvent.click(hamburgerButton!);

    // The mobile nav drawer now has a max height that is not 0 (expanded state).
    await waitFor(() => {
      expect(mobileNavDrawer!.className).not.toMatch(/max-h-0/);
      expect(profileDrawer!.className).toMatch(/max-h-0/); // Profile drawer remains closed
    });
  });

  test('clicking the profile menu button expands the profile drawer', async () => {
    mockLoggedInUser();

    const { container } = render(
      <Nav courses={[{ title: 'Course 1', url: '/course1' }]} />,
    );

    const profileButton = container.querySelector('.nav__profile-menu');
    expect(profileButton).not.toBeNull();

    const mobileNavDrawer = container.querySelector('.nav__links-drawer');
    const profileDrawer = container.querySelector('.nav__profile-drawer');
    expect(mobileNavDrawer).not.toBeNull();
    expect(profileDrawer).not.toBeNull();

    // Initially, both drawers should have a max height of 0 (closed state).
    expect(mobileNavDrawer!.className).toMatch(/max-h-0/);
    expect(profileDrawer!.className).toMatch(/max-h-0/);

    fireEvent.click(profileButton!);

    // The profile drawer now has a max height that is not 0 (expanded state).
    await waitFor(() => {
      expect(profileDrawer!.className).not.toMatch(/max-h-0/);
      expect(mobileNavDrawer!.className).toMatch(/max-h-0/); // Mobile nav drawer remains closed
    });

    vi.clearAllMocks();
  });

  test('clicking outside the nav closes the drawer', async () => {
    const { container } = render(
      <Nav courses={[{ title: 'Course 1', url: '/course1' }]} />,
    );

    const hamburgerButton = container.querySelector('.nav__menu--mobile-tablet');
    expect(hamburgerButton).not.toBeNull();

    const navDrawer = container.querySelector('.nav__links-drawer');
    expect(navDrawer).not.toBeNull();

    fireEvent.click(hamburgerButton!);

    await waitFor(() => {
      expect(navDrawer!.className).not.toMatch(/max-h-0/);
    });

    // Simulate clicking outside the nav drawer
    fireEvent.click(document.body);

    // Ensure the nav drawer is closed
    await waitFor(() => {
      expect(navDrawer!.className).toMatch(/max-h-0/);
    });
  });
});
