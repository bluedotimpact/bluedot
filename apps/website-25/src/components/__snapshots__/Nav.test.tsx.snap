// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nav > renders with courses 1`] = `
<div>
  <nav
    class="nav fixed z-50 w-full container-elevated transition-all duration-300 bg-color-canvas"
  >
    <div
      class="nav__container section-base"
    >
      <div
        class="nav__bar w-full grid grid-cols-2 gap-x-9 lg:grid-cols-[20%_1fr_auto_auto_auto_auto_1fr_20%] items-center grid-rows-[72px] sm:grid-rows-[100px]"
      >
        <a
          class="nav__logo-link shrink-0 w-[200px] col-[1]"
          href="/"
        >
          <img
            alt="BlueDot Impact Logo"
            class="nav__logo h-6 mr-auto transition-all duration-300"
            src="logo.png"
          />
        </a>
        <div
          class="nav-links grid [&>*]:w-fit items-center grid-cols-subgrid grid-rows-subgrid row-[1/3] col-[3/7] nav__links--desktop hidden lg:grid mx-auto"
        >
          <div>
            <button
              class="nav-links__dropdown-button flex items-center gap-2 nav-link-animation"
              type="button"
            >
              Explore
              <svg
                class="size-4 transition-transform"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 9l-7 7-7-7"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
          <div
            class="nav-explore-section__content-wrapper overflow-hidden transition-[max-height,opacity] duration-300 max-h-0 opacity-0 nav-links__explore-section col-span-full row-[2]"
          >
            <div
              class="nav-explore-section___dropdown-content flex flex-col gap-[14px] text-pretty px-6 pb-9"
            >
              <h3
                class="nav-explore-section__dropdown-title font-bold"
              >
                Our courses
              </h3>
              <a
                class="nav-explore-section__dropdown-link"
                href="/course1"
                tabindex="-1"
              >
                Course 1
              </a>
              <a
                class="nav-explore-section__dropdown-link"
                href="/course2"
                tabindex="-1"
              >
                <span
                  class="nav-explore-section__new-badge text-bluedot-normal font-black pr-2"
                >
                  New!
                </span>
                Course 2
              </a>
            </div>
          </div>
          <a
            class="nav-links__link nav-link-animation"
            href="/about"
          >
            About us
          </a>
          <a
            class="nav-links__link nav-link-animation"
            href="/join-us"
          >
            Join us
          </a>
          <a
            class="nav-links__link nav-link-animation"
            href="https://bluedot.org/blog/"
          >
            Blog
          </a>
        </div>
        <div
          class="nav__actions flex gap-space-between ml-auto col-[8]"
        >
          <div
            class="nav__cta-container nav__login--tablet-desktop gap-6 hidden sm:flex"
          >
            <a
              class="cta-button flex items-center justify-center rounded-radius-sm transition-all duration-200 text-sm px-4 py-3 w-fit font-[650] whitespace-nowrap cta-button--primary bg-bluedot-normal link-on-dark nav__primary-cta"
              data-testid="cta-link"
              href="https://donate.stripe.com/5kA3fpgjpdJv6o89AA"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="cta-button__text"
              >
                Support us
              </span>
            </a>
          </div>
          <button
            class="hamburger-button nav__menu--mobile-tablet lg:hidden"
            type="button"
          >
            <svg
              class="hamburger-icon"
              fill="none"
              height="48"
              viewBox="0 0 48 48"
              width="48"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 18H32"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
              <path
                d="M16 24H32"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
              <path
                d="M16 30H32"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="nav__drawer transition-[max-height,opacity] duration-300 relative overflow-hidden px-3 max-h-0 opacity-0"
      >
        <div
          class="nav__drawer-content--mobile-tablet flex flex-col grow font-medium pb-8 pt-2 lg:hidden"
        >
          <div
            class="nav-links grid [&>*]:w-fit items-center grid-cols-1 gap-y-9 nav__links--mobile-tablet flex-col"
          >
            <div>
              <button
                class="nav-links__dropdown-button flex items-center gap-2 nav-link-animation"
                type="button"
              >
                Explore
                <svg
                  class="size-4 transition-transform"
                  fill="none"
                  stroke="currentColor"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 9l-7 7-7-7"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </button>
              <div
                class="nav-explore-section__content-wrapper overflow-hidden transition-[max-height,opacity] duration-300 max-h-0 opacity-0 nav-links__explore-section col-span-full row-[2]"
              >
                <div
                  class="nav-explore-section___dropdown-content flex flex-col gap-[14px] text-pretty px-6 pt-6"
                >
                  <h3
                    class="nav-explore-section__dropdown-title font-bold"
                  >
                    Our courses
                  </h3>
                  <a
                    class="nav-explore-section__dropdown-link"
                    href="/course1"
                    tabindex="-1"
                  >
                    Course 1
                  </a>
                  <a
                    class="nav-explore-section__dropdown-link"
                    href="/course2"
                    tabindex="-1"
                  >
                    <span
                      class="nav-explore-section__new-badge text-bluedot-normal font-black pr-2"
                    >
                      New!
                    </span>
                    Course 2
                  </a>
                </div>
              </div>
            </div>
            <a
              class="nav-links__link nav-link-animation"
              href="/about"
            >
              About us
            </a>
            <a
              class="nav-links__link nav-link-animation"
              href="/join-us"
            >
              Join us
            </a>
            <a
              class="nav-links__link nav-link-animation"
              href="https://bluedot.org/blog/"
            >
              Blog
            </a>
          </div>
          <div
            class="nav__cta-container nav__login--mobile justify-between mt-20 flex sm:hidden"
          >
            <a
              class="cta-button flex items-center justify-center rounded-radius-sm transition-all duration-200 text-sm px-4 py-3 w-fit font-[650] whitespace-nowrap cta-button--primary bg-bluedot-normal link-on-dark nav__primary-cta"
              data-testid="cta-link"
              href="https://donate.stripe.com/5kA3fpgjpdJv6o89AA"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="cta-button__text"
              >
                Support us
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
